DevDelphes - updated output to FCC-EDM (based on new PODIO library)

drasal opened the pull request:
The Delphes simulation module provides now a correct conversion of Delphes internal objects to FCC-EDM objects (based on PODIO) and relates them simultaneously to MC truth information. The definition, which Delphes arrays are automatically translated to FCC-EDM, is given through a GAUDI Pythia+Delphes configuration file. Currently, muons, electrons, charged & neutral hadrons, photons, jets and METS objects are saved together with a full MC particle tree. More details can be found either in a Sim/SimDelphesInterface/src/DelphesSimulation.h header file or config/PythiaDelphes_config.py file.

hegner commented:
test this please

hegner commented:
Could you add a test/test scenario for this adaptor to the pull request? Thanks!

drasal commented:
Dear Benedikt, it has been intensively tested with the real HH data sample.

hegner commented:
@drasal : we would need some test we can run automatically on this comparing delphes and what your translation did. don't want to validate that manually next time we do changes. 

In any case thanks for the pull request. :-) I would have a few general remarks/questions. One is the usage of bare pointers. Please reduce that as much as possible. Why were they needed? Then there are a few strange parts of code where you statically cast iterators. And please instead of uncommenting old code parts, please remove them. I will add a few more concrete remarks later today.

hegner commented on config/PythiaDelphes_config.py:
```diff
@@ -36,17 +50,25 @@
+delphesPhotonsOutArray  ="PhotonEfficiency/photons"
+delphesJetsOutArray     ="JetEnergyScale/jets"
+delphesMETsOutArray     ="MissingET/momentum"
+delphesSHTsOutArray     ="ScalarHT/energy"
```
Why do you define these variables if you use their value only once each? Just simplify.

drasal replied:
Because, the user might want different output from Delphes:
For photons one might get a photon before isolation procedure or after:
PhotonEfficiency/photons or PhotonIsolation/photons 

The same for muons, electrons, etc.

hegner replied:
What I meant is that you can as well do this around line 70ff.

drasal replied:
Ok.

hegner commented on Sim/SimDelphesInterface/src/ParticleStatus.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef INCLUDE_PARTICLE_STATUS_H_
+#define INCLUDE_PARTICLE_STATUS_H_
+
+namespace ParticleStatus {
```
I'd use a typed enum here.

drasal replied:
This is for testing purposes, one should use bit encoder/decoder instead. But yes, I agree enum is probably better.

hegner commented on Sim/SimDelphesInterface/src/ParticleStatus.h:
```diff
@@ -0,0 +1,23 @@
+ *  @author: Z. Drasal (CERN)
+ *
+ */
+#ifndef INCLUDE_PARTICLE_STATUS_H_
```
see
  http://fccsw.web.cern.ch/fccsw/FCCSW/md_doc__cpp_coding_style_guidelines.html#include-guard

drasal replied:
My mistake, I'm sorry.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.h:
```diff
@@ -70,45 +77,66 @@ class DelphesSimulation: public GaudiAlgorithm {
+  DataHandle<fcc::GenJetParticleAssociationCollection>      m_handleRecJetsToMC;      // Relation between jets & MC particle
+
+  // Convert internal Delphes objects to FCC EDM objects - UPDATE
+  void ConvertMCParticles(const TObjArray* Input,    fcc::MCParticleCollection* colMCParticles,
```
passing references would be better. Is there a reason why this wouldn't do?

drasal replied:
Delphes returns a pointer to TObjArray, so I've been working with that. As for the collection, I do create collection dynamically, so then naturally I'm passing them by a pointer. I don't see any possible leak here. If the preferred way is to pass them by reference, it can surely be done.

drasal commented:
Sure, I understand that there needs to be automatic testing. But this time it was unavoidable, due to tests I managed to understand how to make it at all. I will think of automatic testing once it's done, I agree that manual testing is not optimal at all.

You're pointing at:
auto bareJet      = new fcc::BareJet();
well, referencing provided zero values. There might be a trick how to do it without, but I didn't find it. Please, send me a correct example, how to do it, I will try it and correct the code accordingly. 

There is no old code inside. These are the testing outputs.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.h:
```diff
@@ -70,45 +77,66 @@ class DelphesSimulation: public GaudiAlgorithm {
   // Delphes
   Delphes               *m_Delphes ;
   DelphesFactory        *m_DelphesFactory ;
   DelphesExtHepMCReader *m_HepMCReader ;
```
will the Delphes HepMCReader eventually disappear? The idea was to pass HepMC into Delphes from our side... e.g. required if we want to add pile-up ourselves.

drasal replied:
I do delete the pointer in the finalize() method...

hegner replied:
Nah. The point was to have the delphes interface not reading the HepMC file directly, but getting the HepMC event from the FCC framework; in that case the Delphes internal HepMCReader is not needed, right?

drasal replied:
No, it's not needed then.

drasal replied:
So, should I read the HepMC through FCC framework?

hegner replied:
eventually yes.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+  Candidate* cand = nullptr;
+  for(int j=0; j<Input->GetEntries(); j++) {
 
     cand = static_cast<Candidate *>(Input->At(j));
```
I don't understand why you need these static_casts here and at other places. What's the return-type before casting?

drasal replied:
TObject

hegner replied:
I see. If you don't need cand outside the loop, please put the definition inside that scope.

drasal replied:
Alright.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+    }
+    if (barePart!=nullptr) delete barePart;
+
+    // Test print
```
Please remove these lines if they don't serve any purpose any more. Or make it a debug output.

drasal replied:
Ok. I will put them as debug output. Good hint.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+
+    TIter it1(cand->GetCandidates());
+    it1.Reset();
+    while((refCand = static_cast<Candidate*>(it1.Next()))) {
```
here you could just try using iterators; isn't that supported?

drasal replied:
This is the official Delphes code, so I followed it.

hegner replied:
Better follow the standard protocol. Supported here: 
  https://root.cern.ch/doc/master/classTIter.html

you may try whether it already works with
  http://www.cprogramming.com/c++11/c++11-ranged-for-loop.html

drasal replied:
I'm sorry, I didn't get the point. What is wrong about this part of the code. I'm using TIterators... Can you give me a short example.

hegner replied:
see 
http://www.cprogramming.com/tutorial/stl/iterators.html
("The STL approach")
plus the second link above.

drasal replied:
So, you want to use STL iterators instead of ROOT iterators?

hegner replied:
That's just how code could look like, e.g. (untested for syntax errors):
`
for(auto cand = it1.begin(); cand != it1.end(); ++it1 ) {... }
`

drasal replied:
Ok. I know the STL iterators, I just clearly didn't understand what is wrong. So, preferred are STL iterators  ... I will follow them.

jlingema replied:
I think the comment was not to use STL iterators (as you're using a ROOT array here), but rather to use STL-like syntax, i.e. have a look at `Begin` and `End` of ROOT container. The reasoning is that this is likely easier to understand for somebody who is not familiar with ROOT, as it looks more like what you would do with STL containers.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+    Candidate* ref3Cand = nullptr;
+    Candidate* ref4Cand = nullptr;
+
+    TIter it1(cand->GetCandidates());
```
Could you explain why four levels of loops are necessary? The code is not intuitive.

drasal replied:
Well, that's very difficult to answer as some particles in Delphes point directly to the MC particle, some through 1 particle, some through 2 particles and some through 3 particles. That's why, the deep mechanism is hidden within Delphes and I didn't have time to go in detail through the Delphes code. But I understand that this is not a clear solution. I may put there a recursive calling ...

hegner replied:
You mean you "experimentally" found out that 3 levels is enough? You are going through a tree here until you reach the MC particles. Maybe the work of Alice can help.

drasal replied:
No, the Delphes code works until this level of deepness, but one can do it recursively (so something similar what Alice did). In some cases there is direct link, in some cases in goes through related particle, etc. I will think of improving the code.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+      relation.Jet(jet);
+      relation.Particle(colMCParticles->at(id));
+
+      // Test print
```
remove deprecated code. see above.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+    // Test print
+    /*std::cout << std::endl;*/
+
+    if (bareJet!=nullptr) delete bareJet;
```
that makes me believe that you either want to use a unique_ptr or just something on the stack.

drasal replied:
Can you be more specific here. Sorry, my experience is not that large.

jlingema replied:
see http://de.cppreference.com/w/cpp/memory/unique_ptr and https://github.com/HEP-FCC/FCCSW/blob/master/doc/GeneralCppGuidelines.md#poiners-and-references

drasal replied:
Thanks for the comment. I still don't get what is wrong here. Usage of raw pointers is wrong or something more deep?

jlingema replied:
The point is that this is more error prone than using either stack allocation or "smart pointers" (e.g. `unique_ptr`). Since the lifetime of the objects you are creating here is defined by the scope, it makes your life easier not to use pointers. While you now remembered to delete them, somebody modifying the code in the future (copy pasting something) might not.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+    bareJet->P4.Mass  = cand->Mass;
+    jet.Core(*bareJet);
+
+    // Test print
```
see above; remove deprecated code

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -244,84 +265,89 @@ StatusCode DelphesSimulation::execute() {
-    if ( m_outCollections[l].find("mets")     != std::string::npos ) m_recMETHandle.put(recMETs);
-    if ( m_outCollections[l].find("hts")      != std::string::npos ) m_recHTSHandle.put(recHTs);
-  }*/
+  auto genParticles     = new fcc::MCParticleCollection();
```
Not sure if this actually leaks memory, but I think it would be better to use `fcc::MCParticleCollection *genParticles(nullptr);` (same below)

jlingema replied:
Alternatively you could also try to just get the collections from the handles (`auto genParticles = m_handleGenParticles.get();`)

drasal replied:
Ok.

jlingema replied:
Now you can actually use the createAndPut function here :-)

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+      relation.Rec(particle);
+      relation.Sim(colMCParticles->at(id));
+
+      // Test print
```
see above; remove deprecated code

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+    barePart->Vertex.Z = cand->Position.Z();
+    particle.Core(*barePart);
+
+    // Test print
```
see above; remove deprecated code

drasal replied:
I will put debug() in there. Is there several debug levels defined or just one debug level?

jlingema replied:
Only one debug level, see [gaudi-doxygen](http://proj-gaudi.web.cern.ch/proj-gaudi/releases/v26r3/doxygen/d6/d6b/class_gaudi_common.html#ab3f3a23fca6a3375ac5f99c61ae8590d)

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -244,84 +265,89 @@ StatusCode DelphesSimulation::execute() {
+
+  if (m_muonOutArray    ==nullptr) warning () << "Can't save Delphes muon array: "
+                                              << m_DelphesMuonsArrayName
+                                              << " to FCCEDM. Doesn't exist!!!";
```
maybe add a `endmsg()` here?

drasal replied:
Ok. Good point.

jlingema replied:
Comment is still valid.

hegner commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+        int id = ref2Cand->GetUniqueID()-1;
+        if (id<colMCParticles->size()) idRefMCPart.insert(id);
+        else {
+          warning() << "Can't build one of the relations from Photon/NHadron to MC particle!" << std::endl;
```
Why a warning? Is that something one can ignore? Having that fail - is that a feature of the input data or the algorithm?

drasal replied:
This is warning that the algorithm is not working correctly. According to the Delphes code, it should be alright. But if they change it, we have no chance to spot it without a warning line.

hegner replied:
If it fails, it fails. So it should stop and not warn.

drasal replied:
Ok.

jlingema commented:
Regarding the testing output: If this might be needed in the future, why don't you use the debug stream instead of commenting it out?

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.h:
```diff
@@ -3,10 +3,14 @@
-#include "datamodel/MCParticleCollection.h"
 #include "TFile.h"
 #include "FWCore/DataHandle.h"
 #include "HepMC/GenEvent.h"
```
I think several of these could just be forward declarations and then only included in the implementation file.

drasal replied:
Sure, I will test it.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.h:
```diff
@@ -70,45 +77,66 @@ class DelphesSimulation: public GaudiAlgorithm {
-
-  // Output collections to be write out
-  std::vector<std::string> m_outCollections;
+  std::string            m_DelphesCard;
```
This could also be done in another iteration, but since I see that you did some doxygen documentation of the class :-). Could you add some comments for the data and function members? Some names are not completely clear (e.g. the *OutArrays at first glance might look like your output not the Delphes output).

jlingema commented on Generation/src/PythiaInterface.cpp:
```diff
@@ -14,38 +17,48 @@ PythiaInterface::PythiaInterface(const std::string& name, ISvcLocator* svcLoc):
-  declareOutput("hepmc", m_hepmchandle);
+  declareOutput(  "hepmc"   , m_hepmchandle);
+
+  m_nAbort = 0;
```
Please initialise all members in the initialiser list.

jlingema commented on Generation/src/PythiaInterface.cpp:
```diff
@@ -53,55 +66,89 @@ StatusCode PythiaInterface::execute() {
-                << " " << m_pythia->event[i].pz() 
-                << " " << m_pythia->event[i].e()
-                << " " << m_pythia->event[i].m()
+      std::cout << "PythiaInterface : "
```
I guess this should rather go to debug than cout?

jlingema commented on Generation/src/PythiaInterface.cpp:
```diff
@@ -53,55 +66,89 @@ StatusCode PythiaInterface::execute() {
+        daughterIDRange = (*ipart)->end_vertex()->particles_out_size() -1;
+      }
+
+      std::cout << "Pythia HepMC: "
```
Please change this to debug stream..

jlingema commented on Generation/src/PythiaInterface.cpp:
```diff
@@ -53,55 +66,89 @@ StatusCode PythiaInterface::execute() {
+  // Print debug: HepMC event info
+  if (msgLevel() <= MSG::DEBUG) {
+
+    for (HepMC::GenEvent::particle_iterator ipart = theEvent->particles_begin() ;
```
not very important as it is in the debugging clause, but this would be nicer:
```c++
for (auto ipart = theEvent->particles_begin(), end = theEvent->particles_end();
      ipart != end; ++ipart) {
```

jlingema commented on Generation/src/PythiaInterface.h:
```diff
@@ -22,14 +27,19 @@ class PythiaInterface: public GaudiAlgorithm {
   Pythia8::Pythia * m_pythia;
-  /// the name of the Pythia parameter input file
+
+  // Name of Pythia configuration input file
```
Why did you change the doxygen comment to a normal comment? Please re-add the third "/".

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -1,5 +1,33 @@
+#include "TObjArray.h"
+#include "TStopwatch.h"
+
 using namespace std;
```
please re-frain from using `using namespace std`. Instead add explicit statements for the constructs you want to use from `std`. E.g. `using std::cout`.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -21,32 +48,49 @@ GaudiAlgorithm(name, svcLoc) ,
-   declareOutput("recMETs"           , m_recMETHandle);
-   declareOutput("recHTs"            , m_recHTSHandle);
 
    m_stablePartOutArray = nullptr;
```
Please initialise all members in the member initialiser list.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -88,28 +132,28 @@ StatusCode DelphesSimulation::initialize() {
 
-  // Read Delphes configuration card
+  // Read Delphes configuration card (deleted by finalize())
   m_confReader = new ExRootConfReader;
```
Why don't you use smart-pointers here? Same question applies to the other raw pointer members.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -165,67 +209,75 @@ StatusCode DelphesSimulation::execute() {
+    // Print debug: HepMC event info
+    if (msgLevel() <= MSG::DEBUG) {
+
+      for(auto ipart=hepMCEvent->particles_begin(); ipart!=hepMCEvent->particles_end(); ++ipart) {
```
Same comment as for the for-loop in PythiaInterface.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -165,67 +209,75 @@ StatusCode DelphesSimulation::execute() {
+          daughterIDRange = (*ipart)->end_vertex()->particles_out_size() -1;
+        }
+
+        std::cout << "Delphes HepMC: "
```
Probably should also go to debug stream?

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -165,67 +209,75 @@ StatusCode DelphesSimulation::execute() {
+                  << " E: "        << setprecision(2) << setw(9) << (*ipart)->momentum().e()
+                  << " M: "        << setprecision(2) << setw(9) << (*ipart)->momentum().m();
+        if ((*ipart)->production_vertex()!=nullptr) {
+        std::cout << " Vx: "       << setprecision(2) << setw(9) << (*ipart)->production_vertex()->position().x()
```
Increase indentation level.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +400,514 @@ StatusCode DelphesSimulation::finalize() {
-
+  
   // Clear memory
   if (m_HepMCReader) {delete m_HepMCReader; m_HepMCReader = nullptr; } // Releases also the memory allocated by inHepMCFile
```
Add a comparison to nullptr or use smart-pointers.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -343,78 +400,514 @@ StatusCode DelphesSimulation::finalize() {
 
-  Candidate * cand;
-  for(int j = 0; j < Input->GetEntries(); ++j) {
+  Candidate* cand = nullptr;
```
Looks to me that the scope of cand could be decreased to the for-loop below.

drasal commented:
I have updated the code following Joschka's recommendations:

* PythiaInterface: 
 - no raw pointer 
 - added doxygen comments 
 - moved std::cout print to GAUDI standard print outputs (debug() mostly)

* DelphesSimulation 
 - removed some of raw pointers (still some remaining, need to study in more detail Delphes memory management, as the pointers are passed as input parameters to Delphes functions, before I transfer them to smart pointer quantities)
 - added doxygen comments
 - moved std::cout print to GAUDI standard print outputs (debug() mostly)

Remaining issues:
 - Delphes related pointers need to changed to smart pointers (the next step)
 - MC particle tree for some particles broken if LHE file from Madgraph used (under progress)
 - Tag info missing in FCC-EDM & Delphes output (the next step)

hegner commented:
@drasal - the tests fail because of

>  PYTHIA Abort from Pythia::init: Les Houches Event File not found 

can you address that?

drasal commented:
It seems that the automatic test failed because it couldn't find the options/Pythia_LHEinput.cmd, which is in the options directory. Is the directory, where the test is run set to FCCSW directory?

drasal commented:
@hegner When I run it locally at LXPLUS the same procedure is without any problem.

vvolkl commented:
Hi Zbynek,
I had a similar problem restructuring the Generation package. The tests are run from the build directory, but you may try to set `WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}` in `gaudi_add_tests` in `CMakeLists.txt`

drasal commented:
Hi, thanks a lot for the hint. But I checked that in the main CMakeLists.txt it's set like that ...

jlingema commented:
I believe it's due to this line:
https://github.com/drasal/FCCSW/blob/devDelphes/options/Pythia_LHEinput.cmd#L42
The file is not in the repository. Also remember that for the tests you'll need to specify filenames relative to the FCCSW top directory.

drasal commented:
Thank you Joschka! I will fix it.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -341,80 +409,516 @@ StatusCode DelphesSimulation::finalize() {
+    warning() << "Can't build one of the relations from Jet to MC particle!" << std::endl;
+  }
+  // Relation can be found
+  else for (auto itCand=jetPart->GetCandidates()->begin(); itCand!=jetPart->GetCandidates()->end(); ++itCand) {
```
Please add curly braces here for the `else` here.

jlingema replied:
Also please make sure that the for-loops use this slightly nicer syntax that avoids multiple function calls:
```c++
for (auto itCand=jetPart->GetCandidates()->begin(), end = jetPart->GetCandidates()->end(); itCand != end; ++itCand)
```

drasal replied:
GetCandidates returns a pointer, so there is no CPU overhead for direct call. Curl braces will be added.

jlingema commented on Sim/SimDelphesInterface/src/DelphesSimulation.cpp:
```diff
@@ -18,37 +43,54 @@ GaudiAlgorithm(name, svcLoc) ,
+  declareOutput("recJetsToMC"       , m_handleRecJetsToMC);
+
+  // All variables truly initialized in the initialize() method (for pedantic C++ reasons also in a constructor)
+  m_stablePartOutArray = nullptr;
```
Please initialise all members in the member initialiser list.

drasal replied:
They are all initialized in the member initialiser list. Look below.

jlingema replied:
The member initialiser list is above this point in the code :-).. I.e. where you initialise a lot of the other members.

drasal replied:
Ah Ok.

jlingema commented:
Thanks a lot for implementing the changes! Merging this now.
